packages/gatsby/cache-dir/static-query.js

```diff
@@ -2,9 +2,42 @@ import React from "react"
 import PropTypes from "prop-types"
 
 const StaticQueryContext = React.createContext({})
-let StaticQueryServerContext = null
+let _StaticQueryServerContext = null
 if (React.createServerContext) {
-  StaticQueryServerContext = React.createServerContext(`StaticQuery`, {})
+  _StaticQueryServerContext = createServerContext(`StaticQuery`, {})
+}
+
+function createServerOrClientContext(name, defaultValue) {
+  if (React.createServerContext) return createServerContext(name, defaultValue)
+  else return React.createContext(defaultValue)
+}
+
+// Ensure serverContext is not created more than once as React will throw when creating it more than once
+// https://github.com/facebook/react/blob/dd2d6522754f52c70d02c51db25eb7cbd5d1c8eb/packages/react/src/ReactServerContext.js#L101
+function createServerContext(name, defaultValue) {
+  if (!global.__GATSBY_SERVER_CONTEXT__) {
+    global.__GATSBY_SERVER_CONTEXT__ = {}
+  }
+
+  if (!global.__GATSBY_SERVER_CONTEXT__[name]) {
+    global.__GATSBY_SERVER_CONTEXT__[name] = React.createServerContext(
+      name,
+      defaultValue
+    )
+  }
+
+  return global.__GATSBY_SERVER_CONTEXT__[name]
+}
+
+function getClientOrServerContext() {
```

Comment:
Will we need this for more than one variable? Seems like it's worth making the generic.

Something like: `getGatsbyContext("StaticQuery")` that would then look at a map of `global.__GATSBY_SERVER_CONTEXT__` or an internal variable for client side contexts

Reply:
@marvinjude have a look at https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-script/src/collected-scripts-by-page.tsx

Reply:
This can be done when I confirm if current  Implementation is correct